File: backend/api/grades.py
from typing import Annotated
from cachetools import TTLCache
from fastapi import APIRouter, Depends
from fastapi.responses import JSONResponse
from pydantic import BaseModel
from api.auth import User, get_current_user
import db
from fetcher.grade import Grade, GradeFetcher
from fetcher.login import get_session_token
from fetcher.req import verify_session

router = APIRouter()
cache = TTLCache(maxsize=1024, ttl=60*60)

class GradeModel(BaseModel):
    id: int
    value: int
    date: str
    subject_id: int
    entered_by: str
    semester: int
    type: str
    subject_name: str
    short_subject_name: str

class SubjectModel(BaseModel):
    id: int
    name: str
    short_name: str

class SummarizedGrade(BaseModel):
    subject_id: int
    grade_id: int
    value: int
    subject: str

@router.get("/subjects")
def get_subjects(user: Annotated[User, Depends(get_current_user)]) -> list[SubjectModel]:
    return db.get_subjects(user.id)


@router.get("/subjects/{subject_id}")
def get_subjects(subject_id: int,user: Annotated[User, Depends(get_current_user)]) -> SubjectModel:
    data = db.get_subjects(user.id,subject_id)
    if len(data) == 1:
        return data[0]
    elif len(data) == 0:
        return JSONResponse({}, status_code=404)

    # Shouldn't be reachable
    return JSONResponse({}, status_code=400)


class GradeResponse(BaseModel):
    id: int
    value: int
    date: str = "yyyy-mm-dd"
    subject_id: int
    entered_by: str
    semester: int = 1
    type: str 


@router.get("/grades")
def get_grades(user: Annotated[User, Depends(get_current_user)]) -> list[GradeResponse]:
    return db.get_grades(user.id)


@router.get("/summarized_grades")
def get_grades(user: Annotated[User, Depends(get_current_user)]) -> list[SummarizedGrade]:
    return db.get_summarized_grades(user.id)


@router.get("/grades/{subject_id}", )
def get_grades_for_subject(subject_id: int, user: Annotated[User, Depends(get_current_user)]) -> list[GradeResponse]:
    return db.get_grades(user.id, subject_id)


@router.get("/grade/{grade_id}")
def get_grade(grade_id: int, user: Annotated[User, Depends(get_current_user)]) -> GradeResponse:
    return db.get_grade(user.id, grade_id)


@router.get("/average/{subject_id}")
def get_average_grade(subject_id: int, user: Annotated[User, Depends(get_current_user)]):

    x = cache.get(f"{subject_id}_{user.id}")
    if x is not None:
        return {"success": True, "average": x, "subject": subject_id}

    login = db.get_session(user.id)

    if login is None:
        return JSONResponse({"success": False, "message": "No login added yet"}, status_code=400)
    ses = login["ses"]
    if login["ses"] is None or verify_session(login["ses"]):
        ses = get_session_token(login["username"], login["password"])
        db.set_session(user.id, ses)

    fetcher = GradeFetcher(ses)
    avg = fetcher.fetch_avg_grade_for_subject(subject_id=subject_id)
    if avg is None:
        return JSONResponse({"success": False}, status_code=400)

    cache[f"{subject_id}_{user.id}"] = avg
    return {"success": True, "average": avg, "subject": subject_id}


@router.post("/grade/", )
def insert_grade(grade: GradeModel, user: Annotated[User, Depends(get_current_user)]):
    res = db.insert_grade(user.id, Grade(id=grade.id, value=grade.value, subject_id=grade.subject_id, type=grade.type,
                                         entered_by=grade.entered_by, semester=grade.semester, date=grade.date,
                                         short_subject=grade.short_subject_name, subject=grade.subject_name))
    return {"success": res, "message": ""}

File: frontend/src/app/services/grades.ts
import { inject, Injectable } from '@angular/core';
import { Auth } from '../auth';
import { HttpClient } from '@angular/common/http';
import { catchError, map, Observable, of, throwError } from 'rxjs';
import { environment } from '../../environments/environment';

export class Grade {
	id: number;
	value: number;
	date: Date;
	subject_id: number;
	entered_by: string;
	semester: number;
	type: string;

	constructor(
		id: number,
		value: number,
		date: Date,
		subject_id: number,
		entered_by: string,
		semester: number,
		type: string,
	) {
		this.id = id;
		this.value = value;
		this.subject_id = subject_id;
		this.entered_by = entered_by;
		this.semester = semester;
		this.type = type;
		this.date = new Date(date);
	}
	static fromApi(raw: any): Grade {
		return new Grade(
			raw.id,
			raw.value,
			new Date(raw.date),
			raw.subject_id,
			raw.entered_by,
			raw.semester,
			raw.type,
		);
	}
}

export class Subject {
	id: number;
	short_name: string;
	name: string;
	constructor(id: number, short_name: string, name: string) {
		this.id = id;
		this.short_name = short_name;
		this.name = name;
	}
}

export class SummarizedGrade {
	subject_id: number;
	grade_id: number;
	value: number;
	subject: string;

	constructor(subject_id: number, grade_id: number, value: number, subject: string) {
		this.subject_id = subject_id;
		this.grade_id = grade_id;
		this.value = value;
		this.subject = subject;
	}
}

@Injectable({
	providedIn: 'root',
})
export class Grades {
	authService = inject(Auth);
	http = inject(HttpClient);

	getSubjects(): Observable<Subject[]> {
		return this.http.get<Subject[]>(environment.apiUrl + '/subjects').pipe(
			catchError((err) => {
				if (err.status == 401) {
					this.authService.logout();
				}
				return of([]);
			}),
		);
	}

	getSubject(subject_id: string): Observable<Subject | null> {
		return this.http.get<Subject>(environment.apiUrl + '/subjects/' + subject_id).pipe(
			catchError((err) => {
				if (err.status == 401) {
					this.authService.logout();
				}

				return of(null);
			}),
		);
	}

	getGrades(subject_id: number | null): Observable<Grade[]> {
		return this.http.get<any[]>(environment.apiUrl + '/grades/' + (subject_id ?? '')).pipe(
			map((items) => items.map(Grade.fromApi)),
			catchError((err) => {
				if (err.status == 401) {
					this.authService.logout();
				}
				return of([]);
			}),
		);
	}
	getSummarizedGrades(): Observable<SummarizedGrade[]> {
		return this.http.get<SummarizedGrade[]>(environment.apiUrl + '/summarized_grades').pipe(
			catchError((err) => {
				if (err.status == 401) {
					this.authService.logout();
				}
				return of([]);
			}),
		);
	}
	getAverageGrade(subject_id: number): Observable<number> {
		return this.http.get<any[]>(environment.apiUrl + '/average/' + subject_id).pipe(
			map((res: any) => res['average']),
			catchError((err) => {
				if (err.status == 401) {
					this.authService.logout();
				}
				return of(0.0);
			}),
		);
	}

	addLoginInfo(username: string, password: string): Observable<boolean> {
		return this.http
			.post(environment.apiUrl + '/add-login-data', {
				username: username,
				password: password,
			})
			.pipe(
				map((r) => {
					return true;
				}),
				catchError((r) => {
					return of(false);
				}),
			);
	}

	clearData(): Observable<boolean> {
		return this.http.post<{ success: boolean }>(environment.apiUrl + '/clear_data', {}).pipe(
			map((r) => {
				return r.success;
			}),
			catchError((r) => {
				return of(false);
			}),
		);
	}
}
